### DISCUSSION TITLE
How do I set UserDefault values from Android (perhaps from iOS too)

### DISCUSSION THREAD

**Discussion by murasu** [axmolengine/axmol#1495] [Q&A] [answered]
In Cocos2dx, I used the following code to set a user default value in Java
`Cocos2dxHelper.setStringForKey(myKey, myString);`
and I can pick it up in my C++ code with UserDefault::getStringForKey(..).  How do I do the same in Axmol? 

I tried the following:
`AxmolEngine.setStringForKey(myKey, myString);`
But this seems to save it in Android Preferences. I can't pull it up in C++ with UserDefault::getStringForKey(..)

Appreciate any pointers. Thank you!

**Comment by rh101** [comment, marked as answer]
That isn't supported with the current implementation of UserDefault in Axmol, since it is no longer using the Android SharedPreferences implementation.  

You have several options (all very simple to do), such as:

1 - Implement your own UserDefault which would call the AxmolEngine functions for SharedPreferences, then call `UserDefault::setDelegate(UserDefault* delegate)` to set your own implementation.  You would add this custom UserDefault code to your own project.

2 - Add Java code to your application that can call native code in your app, and that in turn will call the current Axmol UserDefault implementation instead, so no need to use Android SharedPreferences etc..

For iOS, it's the same as Android, and you can implement both options, but, unless you specifically need the native key/value storage provided by the iOS/Android, then it may be easier to just go with option 2 for both (on iOS you would add the Objective-C code to call the UserDefault methods).

**Reply by murasu** [reply]
Thank you! This is helpful. On iOS, I am able to call  `UserDefault::getInstance()->setStringForKey` and `UserDefault::getInstance()->getStringForKey` directly in Objective-C. I'll try the options you've suggested for Android.

**Reply by murasu** [reply]
In case anyone is looking for a fix, here's what I ended up doing. I can pick up from Java's system preferences with the help of cocos code that is in Axmol :) 

```
static std::string myGetStringForKey(std::string key, std::string defValue) {
#if (AX_TARGET_PLATFORM == AX_PLATFORM_ANDROID)
        return cocos2d::JniHelper::callStaticStringMethod("org.axmol.lib.AxmolEngine", "getStringForKey", key, defValue);
#else
        return UserDefault::getInstance()->getStringForKey(key, defValue));
#endif
}
```

I can use the same to set something in Axmol/C++ with `setStringForKey()`. This can be picked up at the Java side.

**Comment by antn9x** [comment]
You can use `UserDefault::getInstance()->setStringForKey` and `UserDefault::getInstance()->getStringForKey`.

**Reply by murasu** [reply]
For iOS, yes. Thank you!

**Reply by antn9x** [reply]
This work for both in my game.

**Reply by murasu** [reply]
In the C++ code, yes. My question was how to get/set user defaults in Java.  Adding custom JNI code that calls these functions appears to be an easier solution - until this gets supported in Axmol (hopefully soon).
